connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroDeserializer.scala

```diff
@@ -118,268 +118,225 @@ private[sql] class AvroDeserializer(
-    avroType.getType match {
-      case NULL =>
-        (logicalDataType, catalystType) match {
-          case (_, NullType) => (updater, ordinal, _) =>
-            updater.setNullAt(ordinal)
-          case _ => throw new IncompatibleSchemaException(incompatibleMsg)
-        }
+    (avroType.getType, catalystType) match {
+      case (NULL, NullType) => (updater, ordinal, _) =>
+        updater.setNullAt(ordinal)
+
       // TODO: we can avoid boxing if future version of avro provide primitive accessors.
-      case BOOLEAN =>
-        (logicalDataType, catalystType) match {
-          case (_, BooleanType) => (updater, ordinal, value) =>
-            updater.setBoolean(ordinal, value.asInstanceOf[Boolean])
-          case _ => throw new IncompatibleSchemaException(incompatibleMsg)
-        }
+      case (BOOLEAN, BooleanType) => (updater, ordinal, value) =>
+        updater.setBoolean(ordinal, value.asInstanceOf[Boolean])
+
+      case (INT, IntegerType) => (updater, ordinal, value) =>
+        updater.setInt(ordinal, value.asInstanceOf[Int])
+
+      case (LONG, dt: TimestampType)
+        if preventReadingIncorrectType && realDataType.isInstanceOf[DayTimeIntervalType] =>
+        throw QueryCompilationErrors.avroIncorrectTypeError(toFieldStr(avroPath),
+          toFieldStr(catalystPath), realDataType.catalogString, dt.catalogString, confKey.key)
+
+      case (LONG, dt: TimestampNTZType)
+        if preventReadingIncorrectType && realDataType.isInstanceOf[DayTimeIntervalType] =>
+        throw QueryCompilationErrors.avroIncorrectTypeError(toFieldStr(avroPath),
+          toFieldStr(catalystPath), realDataType.catalogString, dt.catalogString, confKey.key)
+
+      case (LONG, dt: DateType)
+        if preventReadingIncorrectType && realDataType.isInstanceOf[DayTimeIntervalType] =>
+        throw QueryCompilationErrors.avroIncorrectTypeError(toFieldStr(avroPath),
+          toFieldStr(catalystPath), realDataType.catalogString, dt.catalogString, confKey.key)
+
+      case (INT, dt: TimestampType)
+        if preventReadingIncorrectType && realDataType.isInstanceOf[YearMonthIntervalType] =>
+        throw QueryCompilationErrors.avroIncorrectTypeError(toFieldStr(avroPath),
+          toFieldStr(catalystPath), realDataType.catalogString, dt.catalogString, confKey.key)
+
+      case (INT, dt: TimestampNTZType)
+        if preventReadingIncorrectType && realDataType.isInstanceOf[YearMonthIntervalType] =>
+        throw QueryCompilationErrors.avroIncorrectTypeError(toFieldStr(avroPath),
+          toFieldStr(catalystPath), realDataType.catalogString, dt.catalogString, confKey.key)
+
+      case (INT, dt: DateType)
+        if preventReadingIncorrectType && realDataType.isInstanceOf[YearMonthIntervalType] =>
+        throw QueryCompilationErrors.avroIncorrectTypeError(toFieldStr(avroPath),
+          toFieldStr(catalystPath), realDataType.catalogString, dt.catalogString, confKey.key)
+
+      case (INT, DateType) => (updater, ordinal, value) =>
```

Comment:
It seems the logic is also changed. Are you sure where the perf regression comes from? the code structure (pattern match) or some logic?